README.md

# NetBench library

## About
This software, the NetBench library, provides a set of benchmarking libraries
for network quality of service (QoS).  The main target of this software is
the QoS in mobile networks.

## Supported measurement
* ICMP ping (IPv4, IPv6)

## Command line tools

### Run all types of measurement
    $ ./toolset/netbench


ChangeLog
2014-01-22  Hirochika Asai  <asai@scyphus.co.jp>

    * Project started


include/netbench.h
/*_
 * Copyright 2014 Scyphus Solutions Co. Ltd.  All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@scyphus.co.jp>
 */

#ifndef _NETBENCH_H
#define _NETBENCH_H

#include <stdint.h>
#include <netdb.h>

#define USER_AGENT "NetBench/0.1"

/*
 * Measurement results of ping
 */
typedef struct _ping_result_item {
    int stat;
    uint16_t ident;
    double sent;
    double recv;
} nb_ping_result_item_t;
typedef struct _ping_result {
    size_t cnt;
    nb_ping_result_item_t *items;
} nb_ping_result_t;

typedef struct _ping nb_ping_t;

/* Callback function */
typedef void (*nb_ping_cb_f)(nb_ping_t *, int, double);

struct _ping {
    int sock;
    int family;
    nb_ping_cb_f cb;
    void *user;
    nb_ping_result_t *last_results;
    int cancel;
};

/*
 * Traceroute
 */
typedef struct _traceroute_result_item {
    int stat;
    int ttl;
    /*uint16_t sport;*/
    double sent;
    double recv;
    struct sockaddr_storage saddr;
    socklen_t saddrlen;
} nb_traceroute_result_item_t;
typedef struct _traceroute_result {
    size_t cnt;
    size_t cntres;
    nb_traceroute_result_item_t *items;
} nb_traceroute_result_t;
typedef struct _traceroute nb_traceroute_t;
typedef void (*nb_traceroute_cb_f)(nb_traceroute_t *, int,
                                   const struct sockaddr_storage *, socklen_t,
                                   double);
struct _traceroute {
    nb_traceroute_cb_f cb;
    void *user;
    nb_traceroute_result_t *last_results;
    int cancel;
};

/*
 * HTTP (GET)
 */
typedef struct _http_get_result_item {
    double tm;
    off_t tx;
    off_t rx;
    off_t brx;
} nb_http_get_result_item_t;
typedef struct _http_get_result {
    size_t cnt;
    size_t cntres;
    nb_http_get_result_item_t *items;
    off_t hlen;
    off_t clen;
    int mss;
} nb_http_get_result_t;
typedef struct _http_get nb_http_get_t;
typedef void (*nb_http_get_cb_f)(nb_http_get_t *, off_t, off_t, double, double,
                                 off_t, off_t);
struct _http_get {
    nb_http_get_cb_f cb;
    double cbfreq;
    void *user;
    char *mid;
    nb_http_get_result_t *last_result;
    int cancel;
};

/*
 * HTTP (POST)
 */
typedef struct _http_post_result_item {
    double tm;
    off_t tx;
    off_t rx;
    off_t btx;                 /* Buffered TX */
} nb_http_post_result_item_t;
typedef struct _http_post_result {
    size_t cnt;
    size_t cntres;
    nb_http_post_result_item_t *items;
    off_t hlen;
    off_t clen;
    int mss;
} nb_http_post_result_t;
typedef struct _http_post nb_http_post_t;
typedef void (*nb_http_post_cb_f)(nb_http_post_t *, off_t, off_t, double,
                                  double, off_t, off_t);
struct _http_post {
    nb_http_post_cb_f cb;
    double cbfreq;
    char *mid;
    void *user;
    nb_http_post_result_t *last_result;
    int cancel;
};

/*
 * Parsed URL
 */
typedef struct _parsed_url {
    char *scheme;               /* mandatory */
    char *host;                 /* mandatory */
    char *port;                 /* optional */
    char *path;                 /* optional */
    char *query;                /* optional */
    char *fragment;             /* optional */
    char *username;             /* optional */
    char *password;             /* optional */
} nb_parsed_url_t;

/*
 * HTTP header
 */
typedef struct _http_header_attr {
    char *key;
    char *value;
} nb_http_header_attr_t;
typedef struct _http_header_attr_list nb_http_header_attr_list_t;
struct _http_header_attr_list {
    nb_http_header_attr_t *attr;
    nb_http_header_attr_list_t *head;
    nb_http_header_attr_list_t *prev;
    nb_http_header_attr_list_t *next;
};
typedef struct _http_header {
    /* Method */
    char *method;
    /* Uri */
    char *uri;
    /* Version */
    char *version;
    /* Attributes */
    nb_http_header_attr_list_t *attrs;
} nb_http_header_t;


#ifdef __cplusplus
extern "C" {
#endif

    /* Generic functions */
    double nb_microtime(void);
    uint16_t nb_checksum(const uint8_t *, size_t);
    nb_parsed_url_t * nb_parse_url(const char *);
    void nb_parsed_url_free(nb_parsed_url_t *);
    nb_http_header_t * nb_parse_http_header(const char *, size_t);
    void nb_http_header_delete(nb_http_header_t *);
    off_t nb_http_header_get_content_length(nb_http_header_t *);
    int nb_http_get(const char *url, char **, off_t *);
    int
    nb_http_post(const char *, const char *, const char *, size_t, char **,
                 off_t *);

    /* Ping */
    nb_ping_t * nb_ping_open(int);
    int nb_ping_set_callback(nb_ping_t *, nb_ping_cb_f, void *);
    int nb_ping_exec(nb_ping_t *, const char *, size_t, int, double, double);
    void nb_ping_close(nb_ping_t *);

    /* Traceroute */
    nb_traceroute_t * nb_traceroute_new(void);
    int
    nb_traceroute_set_callback(nb_traceroute_t *, nb_traceroute_cb_f, void *);
    int nb_traceroute_exec(nb_traceroute_t *, const char *, int, int, double);
    void nb_traceroute_delete(nb_traceroute_t *);

    /* HTTP (GET) */
    nb_http_get_t * nb_http_get_new(const char *);
    int
    nb_http_get_set_callback(nb_http_get_t *, nb_http_get_cb_f, double, void *);
    int nb_http_get_exec(nb_http_get_t *, const char *, int, double);
    void nb_http_get_delete(nb_http_get_t *);

    /* HTTP (POST) */
    nb_http_post_t * nb_http_post_new(const char *);
    int
    nb_http_post_set_callback(nb_http_post_t *, nb_http_post_cb_f, double,
                              void *);
    int nb_http_post_exec(nb_http_post_t *, const char *, int, off_t, double);
    void nb_http_post_delete(nb_http_post_t *);


#ifdef __cplusplus
}
#endif

#endif /* _NETBENCH_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


libnb/traceroute.c
/*_
 * Copyright 2014 Scyphus Solutions Co. Ltd.  All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@scyphus.co.jp>
 */

#include "config.h"
#include "netbench.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <unistd.h>

#define BUFFER_SIZE                     65536
#define TRACEROUTE_USLEEP               1000

struct icmp_hdr {
    uint8_t type;
    uint8_t code;
    uint16_t checksum;
    uint32_t pad;
    // data...
} __attribute__ ((packed));

/*
 * Open a traceroute socket
 */
nb_traceroute_t *
nb_traceroute_new(void)
{
    nb_traceroute_t *obj;

    /* Allocate memory for the ping4 object */
    obj = malloc(sizeof(nb_traceroute_t));
    if ( NULL == obj ) {
        return NULL;
    }
    obj->cb = NULL;
    obj->last_results = NULL;
    obj->cancel = 0;

    return obj;
}

/*
 * Set a callback function
 */
int
nb_traceroute_set_callback(nb_traceroute_t *obj, nb_traceroute_cb_f cbfunc,
                           void *user)
{
    /* Set the callback function */
    obj->cb = cbfunc;
    obj->user = user;

    return 0;
}

/*
 * Receive an ICMP time exceeded (FIXME)
 */
static int
_icmp4_recv(int sock, struct sockaddr_storage *saddr, socklen_t *saddrlen,
            double *t0)
{
    uint8_t buf[BUFFER_SIZE];
    ssize_t nr;

    *saddrlen = sizeof(struct sockaddr_storage);
    nr = recvfrom(sock, buf, BUFFER_SIZE, 0, (struct sockaddr *)saddr,
                  saddrlen);
    *t0 = nb_microtime();
    if ( nr < 0 ) {
        /* Read nothing */
        return -1;
    }
    if ( nr < 24 ) {
        return -1;
    }
    if ( 0x45 != buf[0] ) {
        /* IP */
        return -1;
    }
    if ( 0x01 != buf[9] ) {
        /* Must be ICMP */
        return -1;
    }
    if ( 0x0b != buf[20] && 0x03 != buf[20] ) {
        /* Must be time exceed or destination unreachable */
        return -1;
    }

    /* Check the source address */
    if ( AF_INET != saddr->ss_family ) {
        return -1;
    }

    return 0;
}
static int
_icmp6_recv(int sock, struct sockaddr_storage *saddr, socklen_t *saddrlen,
            double *t0)
{
    uint8_t buf[BUFFER_SIZE];
    ssize_t nr;

    nr = recvfrom(sock, buf, BUFFER_SIZE, 0, (struct sockaddr *)saddr,
                  saddrlen);
    *t0 = nb_microtime();
    if ( nr < 0 ) {
        /* Read nothing */
        return -1;
    }
    if ( nr < 24 ) {
        return -1;
    }
    if ( 0x01 != buf[0] && 0x03 != buf[0] ) {
        /* Must be Destination unreachable or Time exceeded */
        return -1;
    }

    /* Check the source address */
    if ( AF_INET6 != saddr->ss_family ) {
        return -1;
    }

    return 0;
}

/*
 * Execute traceroute
 */
int
nb_traceroute_exec(nb_traceroute_t *obj, const char *target, int family,
                   int maxttl, double timeout)
{
    int sock;
    int icmpsock;
    int error;
    int ttl;
    int opt;
    int i;
    double t0;
    double t1;
    struct addrinfo ai;
    struct addrinfo hints, *ai0, *ressave;
    int pktsize = 40;
    const char *service = "33434";
    struct timeval recv_tv;
    uint8_t buf[BUFFER_SIZE];
    ssize_t sz;
    int ret;
    struct sockaddr_storage saddr;
    socklen_t saddrlen;
    nb_traceroute_result_t *result;

    /* Open an ICMP socket */
    if ( AF_INET == family ) {
        icmpsock = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
    } else if ( AF_INET6 == family ) {
        icmpsock = socket(AF_INET6, SOCK_DGRAM, IPPROTO_ICMPV6);
    } else {
        return -1;
    }
    if ( icmpsock < 0 ) {
        /* Cannot open ICMP socket */
        return -1;
    }

    /* Set timeout */
    recv_tv.tv_sec = (int)timeout;
    recv_tv.tv_usec = (int)((timeout - (int)timeout) * 1000000);
    setsockopt(icmpsock, SOL_SOCKET, SO_RCVTIMEO, &recv_tv, sizeof(recv_tv));

    /* Prepare buffer */
    for ( i = 0; i < pktsize; i++ ) {
        buf[i] = i & 0xff;
    }

    /* Open an UDP socket */
    sock = -1;
    bzero(&ai, sizeof(struct addrinfo));
    /* Setup hints to get address info for a UDP socket */
    bzero(&hints, sizeof(struct addrinfo));
    hints.ai_family = family;
    hints.ai_socktype = SOCK_DGRAM;
    /* Obtain address info of the target */
    error = getaddrinfo(target, service, &hints, &ressave);
    if ( 0 != error ) {
        /* Error on getaddrinfo */
        close(icmpsock);
        return -1;
    }
    /* Search and open the first available socket */
    sock = -1;
    for ( ai0 = ressave; NULL != ai0; ai0 = ai0->ai_next ) {
        /* Open a socket */
        sock = socket(ai0->ai_family, ai0->ai_socktype, ai0->ai_protocol);
        if ( sock < 0 ) {
            /* Ignore unavailable sockets */
            continue;
        }
        /* Succeed */
        memcpy(&ai, ai0, sizeof(struct addrinfo));
        break;
    }
    /* Check the socket */
    if( sock < 0 ){
        freeaddrinfo(ressave);
        close(icmpsock);
        return -1;
    }

    /* Allocate for the results */
    result = malloc(sizeof(nb_traceroute_result_t));
    if ( NULL == result ) {
        return -1;
    }
    result->cnt = 0;
    result->cntres = maxttl + 1;
    result->items = malloc(sizeof(nb_traceroute_result_item_t) * result->cntres);
    if ( NULL == result->items ) {
        free(result);
        return -1;
    }

    /* For each TTL */
    for ( ttl = 1; ttl <= maxttl; ttl++ ) {
        if ( obj->cancel ) {
            break;
        }

        result->cnt = ttl;
        result->items[ttl - 1].ttl = ttl;
        result->items[ttl - 1].stat = -1;

        opt = ttl;
        if ( AF_INET6 == family ) {
            error = setsockopt(sock, IPPROTO_IPV6, IPV6_UNICAST_HOPS, &opt,
                               sizeof(int));
        } else {
            error = setsockopt(sock, IPPROTO_IP, IP_TTL, &opt, sizeof(int));
        }
        if ( 0 != error ) {
            printf(" *");
            continue;
        }
        usleep(TRACEROUTE_USLEEP);

        t0 = nb_microtime();
        sz = sendto(sock, buf, pktsize, 0, ai.ai_addr, ai.ai_addrlen);
        if ( sz < 0 ) {
            continue;
        }
        result->items[ttl - 1].stat = 0;
        result->items[ttl - 1].sent = t0;

        if ( AF_INET == family ) {
            ret = _icmp4_recv(icmpsock, &saddr, &saddrlen, &t1);
            if ( ret < 0 ) {
                continue;
            }
            (void)memcpy(&result->items[ttl - 1].saddr, &saddr,
                         sizeof(struct sockaddr_storage));
            if ( NULL != obj->cb ) {
                obj->cb(obj, ttl, &saddr, saddrlen,
                        t1 - result->items[ttl - 1].sent);
            }
            result->items[ttl - 1].stat++;
            result->items[ttl - 1].recv = t1;
            /* Check destination */
            if ( ((struct sockaddr_in *)ai.ai_addr)->sin_addr.s_addr
                 == ((struct sockaddr_in *)&saddr)->sin_addr.s_addr ) {
                break;
            }
        } else if ( AF_INET6 == family ) {
            ret = _icmp6_recv(icmpsock, &saddr, &saddrlen, &t1);
            if ( ret < 0 ) {
                continue;
            }
            (void)memcpy(&result->items[ttl - 1].saddr, &saddr,
                         sizeof(struct sockaddr_storage));
            if ( NULL != obj->cb ) {
                obj->cb(obj, ttl, &saddr, saddrlen,
                        t1 - result->items[ttl - 1].sent);
            }
            result->items[ttl - 1].stat++;
            result->items[ttl - 1].recv = t1;
            /* Check destination */
            if ( memcmp(((struct sockaddr_in6 *)ai.ai_addr)->sin6_addr.s6_addr,
                        ((struct sockaddr_in6 *)&saddr)->sin6_addr.s6_addr,
                        16) == 0 ) {
                break;
            }
        }
    }

    /* Close */
    freeaddrinfo(ressave);
    close(icmpsock);
    close(sock);

    /* Free the result */
    if ( NULL != obj->last_results ) {
        free(obj->last_results);
    }
    obj->last_results = result;

    return 0;
}

/*
 * Close the traceroute socket
 */
void
nb_traceroute_delete(nb_traceroute_t *obj)
{
    if ( NULL != obj->last_results ) {
        free(obj->last_results);
    }
    free(obj);
}


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


libnb/ping.c
/*_
 * Copyright 2014 Scyphus Solutions Co. Ltd.  All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@scyphus.co.jp>
 */

#include "config.h"
#include "netbench.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <unistd.h>
#include <poll.h>
#include <errno.h>

#define BUFFER_SIZE                     65536
#define ICMP_TYPE_ECHO_REQUEST          8
#define ICMP_TYPE_ECHO_REPLY            0
#define ICMPV6_TYPE_ECHO_REQUEST        128
#define ICMPV6_TYPE_ECHO_REPLY          129

struct ip_hdr {
    uint8_t verlen;
    uint8_t diffserv;
    uint16_t total_len;
    uint16_t ident;
    uint16_t offset;
    uint8_t  ttl;
    uint8_t  protocol;
    uint16_t headerChecksum;
    uint8_t sourceAddress[4];
    uint8_t destinationAddress[4];
    // options...
    // data...
} __attribute__ ((packed));

struct icmp_hdr {
    uint8_t type;
    uint8_t code;
    uint16_t checksum;
    uint16_t ident;
    uint16_t seq;
    // data...
} __attribute__ ((packed));

struct icmp6_hdr {
    uint8_t type;
    uint8_t code;
    uint16_t checksum;
    // data...
} __attribute__ ((packed));

/* Prototype declarations */
static int
_ping_send(nb_ping_t *, struct addrinfo *, uint16_t, uint16_t, size_t,
           double *);
static int
_ping_recv(nb_ping_t *, struct addrinfo *, uint16_t *, double *,
           nb_ping_result_t *);


/*
 * Send an ICMP echo request
 */
static int
_ping_send(nb_ping_t *obj, struct addrinfo *dai, uint16_t ident, uint16_t seq,
           size_t sz, double *tm)
{
    ssize_t ret;
    uint8_t buf[BUFFER_SIZE];
    struct icmp_hdr *icmp;
    size_t pktsize;
    size_t i;

    /* Compute packet size */
    pktsize = sizeof(struct icmp_hdr) + sz;
    if ( pktsize > BUFFER_SIZE ) {
        return -1;
    }

    /* Build an ICMP packet */
    switch ( obj->family ) {
    case AF_INET:
        icmp = (struct icmp_hdr *)buf;
        icmp->type = ICMP_TYPE_ECHO_REQUEST;
        icmp->code = 0;
        icmp->checksum = 0;
        icmp->ident = htons(ident);
        icmp->seq = htons(seq);
        break;
    case AF_INET6:
        icmp = (struct icmp_hdr *)buf;
        icmp->type = ICMPV6_TYPE_ECHO_REQUEST;
        icmp->code = 0;
        icmp->checksum = 0;
        icmp->ident = htons(ident);
        icmp->seq = htons(seq);
        break;
    default:
        return -1;
    }


    /* Fill with values */
    for ( i = sizeof(struct icmp_hdr); i < pktsize; i++ ) {
        buf[i] = i % 0xff;
    }
    icmp->checksum = nb_checksum(buf, pktsize);

    *tm = nb_microtime();
    ret = sendto(obj->sock, buf, pktsize, 0, (struct sockaddr *)dai->ai_addr,
                 dai->ai_addrlen);
    if ( ret < 0 ) {
        /* Failed to send the packet */
        return -1;
    }

    return 0;
}

/*
 * Receive an ICMP echo reply
 */
static int
_ping_recv(nb_ping_t *obj, struct addrinfo *dai, uint16_t *seq, double *tm,
           nb_ping_result_t *res)
{
    uint8_t buf[BUFFER_SIZE];
    struct sockaddr_storage saddr;
    socklen_t saddrlen;
    ssize_t nr;
    struct icmp_hdr *ricmp;
    struct ip_hdr *rip;
    int iphdrlen;
    struct sockaddr_in *sin4;
    struct sockaddr_in6 *sin6;

    saddrlen = sizeof(saddr);
    nr = recvfrom(obj->sock, buf, BUFFER_SIZE, 0, (struct sockaddr *)&saddr,
                  &saddrlen);
    *tm = nb_microtime();
    if ( nr < 0 ) {
        /* Read nothing */
        return -1;
    }

    /* Check the source address */
    if ( dai->ai_family != saddr.ss_family ) {
        return -1;
    }
    if ( AF_INET == dai->ai_family ) {
        sin4 = (struct sockaddr_in *)dai->ai_addr;
        if ( ((struct sockaddr_in *)&saddr)->sin_addr.s_addr
             != sin4->sin_addr.s_addr ) {
            return -1;
        }

        /* Check it */
        if ( nr < sizeof(struct ip_hdr) ) {
            return -1;
        }

        /* IP header */
        rip = (struct ip_hdr *)buf;
        iphdrlen = rip->verlen & 0xf;
        if ( nr < 4 * iphdrlen + sizeof(struct icmp_hdr) ) {
            return -1;
        }
        /* Skip IP header */
        ricmp = (struct icmp_hdr *)(buf + 4 * iphdrlen);
        if ( ICMP_TYPE_ECHO_REPLY != ricmp->type || 0 != ricmp->code ) {
            /* Error */
            return -1;
        }
        *seq = ntohs(ricmp->seq);
        if ( *seq >= res->cnt ) {
            /* Invalid sequence */
            return -1;
        }
        if ( ntohs(ricmp->ident) != res->items[*seq].ident ) {
            /* Invalid identifier */
            return -1;
        }

    } else if ( AF_INET6 == dai->ai_family ) {
        sin6 = (struct sockaddr_in6 *)dai->ai_addr;
        if ( 0 != memcmp(((struct sockaddr_in6 *)&saddr)->sin6_addr.s6_addr,
                         sin6->sin6_addr.s6_addr,
                         sizeof(sin6->sin6_addr.s6_addr)) ) {
        }

        /* Check it */
        if ( nr < sizeof(struct icmp_hdr) ) {
            return -1;
        }

        /* ICMPv6 */
        ricmp = (struct icmp_hdr *)buf;
        if ( ICMPV6_TYPE_ECHO_REPLY != ricmp->type || 0 != ricmp->code ) {
            /* Error */
            return -1;
        }
        *seq = ntohs(ricmp->seq);
        if ( *seq >= res->cnt ) {
            /* Invalid sequence */
            return -1;
        }
        if ( ntohs(ricmp->ident) != res->items[*seq].ident ) {
            /* Invalid identifier */
            return -1;
        }
    } else {
        return -1;
    }

    return 0;
}



/*
 * Open a ping4 socket
 */
nb_ping_t *
nb_ping_open(int family)
{
    nb_ping_t *obj;

    /* Allocate memory for the ping4 object */
    obj = malloc(sizeof(nb_ping_t));
    if ( NULL == obj ) {
        return NULL;
    }

    /* Open the socket */
    switch ( family ) {
    case AF_INET:
#if TARGET_FREEBSD || TARGET_NETBSD || TARGET_LINUX
        obj->sock = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
#else
        obj->sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
#endif
        break;
    case AF_INET6:
#if TARGET_FREEBSD || TARGET_NETBSD || TARGET_LINUX
        obj->sock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
#else
        obj->sock = socket(AF_INET6, SOCK_DGRAM, IPPROTO_ICMPV6);
#endif
        break;
    default:
        /* Invalid family */
        free(obj);
        return NULL;
    }
    if ( obj->sock < 0 ) {
        /* Cannot open the ICMP socket */
        free(obj);
        return NULL;
    }
    obj->family = family;
    obj->cb = NULL;
    obj->last_results = NULL;
    obj->cancel = 0;

    return obj;
}

/*
 * Set a callback function
 */
int
nb_ping_set_callback(nb_ping_t *obj, nb_ping_cb_f cbfunc, void *user)
{
    /* Set the callback function */
    obj->cb = cbfunc;
    obj->user = user;

    return 0;
}

/*
 * Send a ping packet and receive its reply
 */
int
nb_ping_exec(nb_ping_t *obj, const char *target, size_t sz, int n,
             double interval, double timeout)
{
    struct addrinfo *ai;
    struct addrinfo hints;
    struct addrinfo *ressave;
    int i;
    int err;
    int done;
    int nsent;
    int nrecv;
    double t0;
    double t1;
    double tm;
    uint16_t ident;
    uint16_t seq;
    struct pollfd fds[1];
    int events;
    double gto;
    nb_ping_result_t *res;

    /* Setup ai and hints to get address info */
    bzero(&hints, sizeof(struct addrinfo));
    hints.ai_family = obj->family;
    hints.ai_socktype = SOCK_DGRAM;
    err = getaddrinfo(target, NULL, &hints, &ressave);
    if ( 0 != err ) {
        /* Cannot resolve the target host */
        return -1;
    }

    /* Save the first addrinfo */
    if ( NULL == ressave ) {
        /* Error if the first addrinfo is NULL */
        freeaddrinfo(ressave);
        return -1;
    }
    ai = ressave;

    /* Allocate for the results */
    res = malloc(sizeof(nb_ping_result_t));
    if ( NULL == res ) {
        /* Free the returned addrinfo */
        freeaddrinfo(ressave);
        return -1;
    }
    res->cnt = n;
    res->items = malloc(sizeof(nb_ping_result_item_t) * res->cnt);
    if ( NULL == res->items ) {
        free(res);
        /* Free the returned addrinfo */
        freeaddrinfo(ressave);
        return -1;
    }
    for ( i = 0; i < n; i++ ){
        res->items[i].stat = -1;
    }

    /* Obtain the started time */
    t0 = nb_microtime();

    done = 0;
    nsent = 0;
    nrecv = 0;
    while ( !obj->cancel && !done ) {
        /* Obtain the current time */
        t1 = nb_microtime();

        if ( nsent < n && interval * nsent < t1 - t0 ) {
            /* FIXME? */
            ident = random();
            err = _ping_send(obj, ai, ident, nsent, sz, &tm);
            if ( 0 != err ) {
                obj->cancel = 1;
                continue;
            }

            /* Set the result item */
            res->items[nsent].stat = 0;
            res->items[nsent].ident = ident;
            res->items[nsent].sent = tm;
            nsent++;
        }

        /* Update the current time */
        t1 = nb_microtime();

        /* Calculate the timeout for polling */
        if ( nsent < n ) {
            /* Check the timeout to the next packet */
            gto = interval * nsent - (t1 - t0);
            if ( gto <= 0.0 ) {
                /* Send again */
                continue;
            }
        } else {
            /* Check the timeout to the end of the measurement */
            gto = interval * nsent - (t1 - t0) + timeout;
            if ( gto <= 0.0 ) {
                gto = 0.0;
            }
        }

        /* Poll */
        fds[0].fd = obj->sock;
        fds[0].events = POLLIN;
        fds[0].revents = 0;
        events = poll(fds, 1, (int)(gto * 1000));

        if ( events < 0 ) {
            if ( EINTR == errno ) {
                /* Interrupt */
                continue;
            } else {
                /* Other errors */
                obj->cancel = 1;
                continue;
            }
        } else if ( 0 == events ) {
            /* Timeout */
            t1 = nb_microtime();
            if ( interval * n + timeout < (t1 - t0) ) {
                /* Reached the timeout */
                done = 1;
            }
            continue;
        } else {
            if ( fds[0].revents & (POLLERR | POLLHUP | POLLNVAL) ) {
                /* Error */
                obj->cancel = 1;
                continue;
            } else {
                if ( fds[0].revents & POLLIN ) {
                    /* Received */
                    err = _ping_recv(obj, ai, &seq, &tm, res);
                    if ( 0 != err ) {
                        continue;
                    }

                    /* Call the callback function */
                    if ( NULL != obj->cb ) {
                        obj->cb(obj, seq, tm - res->items[seq].sent);
                    }

                    /* Set results */
                    res->items[seq].stat++;
                    res->items[seq].recv = tm;
                    nrecv++;
                    if ( nrecv >= n ) {
                        done = 1;
                    }
                }
            }
        }
    }

    /* Free the result */
    if ( NULL != obj->last_results ) {
        free(obj->last_results->items);
        free(obj->last_results);
    }
    obj->last_results = res;

    /* Free the returned addrinfo */
    freeaddrinfo(ressave);

    return 0;
}

/*
 * Close the ping socket
 */
void
nb_ping_close(nb_ping_t *obj)
{
    (void)close(obj->sock);
    if ( NULL != obj->last_results ) {
        free(obj->last_results->items);
        free(obj->last_results);
    }
    free(obj);
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */
